micro_sam/training/training.py

```diff
@@ -0,0 +1,117 @@
+) -> None:
+    """Run training for a SAM model.
+
+    Args:
+        name: The name of the model to be trained.
+            The checkpoint and logs wil have this name.
+        model_type: The type of the SAM model.
+        train_loader: The dataloader for training.
+        val_loader: The dataloader for validation.
+        n_epochs: The number of epochs to train for.
+        early_stopping: Enable early stopping after this number of epochs
+            without improvement.
+        n_objects_per_batch: The number of objects per batch used to compute
+            the loss for interative segmentation. If None all objects will be used,
+            if given objects will be randomly sub-sampled.
+        checkpoint_path: Path to checkpoint for initializing the SAM model.
+        with_segmentation_decoder: Whether to train additional UNETR decoder
+            for automatic instance segmentation.
+        freeze: Specify parts of the model that should be frozen, namely: image_encoder, prompt_encoder and mask_decoder
+            By default nothing is frozen and the full model is updated.
+        device: The device to use for training.
+        lr: The learning rate.
+        scheduler: The learning rate scheduler. By default ReduceLROnPlateau is used.
+        n_sub_iteration: The number of iterative prompts per training iteration.
+        save_root: Optional root directory for saving the checkpoints and logs.
+            If not given the current working directory is used.
+    """
+    # TODO check the data loaders!
+
+    device = get_device(device)
+
+    # Get the trainable segment anything model.
+    model, state = get_trainable_sam_model(
+        model_type=model_type, device=device, freeze=freeze,
+        checkpoint_path=checkpoint_path, return_state=True,
+    )
+
+    # This class creates all the training data for a batch (inputs, prompts and labels).
+    convert_inputs = ConvertToSamInputs(transform=model.transform, box_distortion_factor=0.025)
+
+    # Create the UNETR decoder (if train with it) and the optimizer.
+    if with_segmentation_decoder:
+
+        # Get the UNETR.
+        unetr = get_unetr(
+            image_encoder=model.sam.image_encoder,
+            decoder_state=state.get("decoder_state", None),
+            device=device,
+        )
+
+        # Get the parameters for SAM and the decoder from UNETR.
+        joint_model_params = [params for params in model.parameters()]  # sam parameters
+        for param_name, params in unetr.named_parameters():  # unetr's decoder parameters
+            if not param_name.startswith("encoder"):
+                joint_model_params.append(params)
+
+        optimizer = torch.optim.Adam(joint_model_params, lr=lr)
+
+    else:
+        optimizer = torch.optim.Adam(model.parameters(), lr=lr)
+
+    if scheduler is None:
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer, mode="min", factor=0.9, patience=3, verbose=True
```

Comment:
I think these parameters can be exposed as an optional parameter, where the defaults could be the values as mentioned here, and can be updated using something like `scheduler_kwargs`.

Reply:
I think this makes it too complex. The scheduler can already be passed, so I think it's better not to introduce any `scheduler_kwargs` here. If a different scheduling behavior is wanted the user can pass their own scheduler.

Reply:
Oh yes, you are right. Didn't spot it hehe. Then that should do the trick!